### PR TITLE
KAFKA-2614; No more clients can connect after `TooManyConnectionsException` threshold (max.connections.per.ip) is reached

### PR DESCRIPTION
**Pull request by ijuma**
- Call `ConnectionQuotas.decr` when calling `Selector.close` and when disconnections happen.
- Expand `SocketServerTest` to test for this and to close sockets.
- Refactor and clean-up `SocketServer` and `Acceptor` to make the code easier to understand.

### CODE REVIEW

**Comment by ijuma** [comment]
Tests passed locally.

**Comment by ijuma** [comment]
System tests passed: http://jenkins.confluent.io/job/kafka_system_tests_branch_builder/97/

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 1]
```diff
@@ -353,18 +349,40 @@ private[kafka] class Acceptor(val host: String,
...
-                               val channelConfigs: java.util.Map[String, Object],
-                               val metrics: Metrics) extends AbstractServerThread(connectionQuotas) with KafkaMetricsGroup {
+                               connectionsMaxIdleMs: Long,
+                               protocol: SecurityProtocol,
+                               channelConfigs: java.util.Map[String, Object],
+                               metrics: Metrics) extends AbstractServerThread(connectionQuotas) with KafkaMetricsGroup {
+
+  private object ConnectionId {
+
+    private def parseHostPort(s: String): Option[(String, Int)] = s.split(":") match {
```
Comment (junrao) [inline core/src/main/scala/kafka/network/SocketServer.scala]: Does this work for ipv6? We probably should use the parser in BrokerEndPoint.

Comment (ijuma) [inline core/src/main/scala/kafka/network/SocketServer.scala]: Good point, fixed as you suggested.

File: clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java [hunk 2]
```diff
@@ -108,14 +109,21 @@ public boolean hasSend() {
...
+     * connected address after the socket is closed.
+     */
+    public InetAddress socketAddress() {
+        return transportLayer.socketChannel().socket().getInetAddress();
+    }
+
     public String socketDescription() {
         Socket socket = transportLayer.socketChannel().socket();
-        if (socket == null)
-            return "[unconnected socket]";
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java:114]: As far as I can see `socket` can never be null here.

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 3]
```diff
@@ -77,30 +74,47 @@ class SocketServer(val config: KafkaConfig, val metrics: Metrics, val time: Time
...
+          processors.slice(processorBeginIndex, processorEndIndex), connectionQuotas)
         acceptors.put(endpoint, acceptor)
-        Utils.newThread("kafka-socket-acceptor-%s-%d".format(endpoint.protocolType.toString, endpoint.port), acceptor, false).start()
-        acceptor.awaitStartup
-        processorBeginIndex += numProcessorThreads
-      })
+        Utils.newThread("kafka-socket-acceptor-%s-%d".format(protocol.toString, endpoint.port), acceptor, false).start()
+        acceptor.awaitStartup()
+
+        processorBeginIndex = processorEndIndex
```
Comment (ijuma) [inline core/src/main/scala/kafka/network/SocketServer.scala:116]: This change is not actually related to the issue, but I found the SocketServer code a bit messy and hard to understand, so I cleaned it up while debugging. If you'd prefer these changes to be done separately, let me know.

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 4]
```diff
@@ -353,18 +349,32 @@ private[kafka] class Acceptor(val host: String,
...
+          ConnectionId(localHost, localPort, remoteHost, remotePort)
+        }
+      }
+      case _ => None
+    }
+  }
+
+  private case class ConnectionId(localHost: String, localPort: Int, remoteHost: String, remotePort: Int) {
+    override def toString: String = s"$localHost:$localPort-$remoteHost:$remotePort"
+  }
```
Comment (ijuma) [inline core/src/main/scala/kafka/network/SocketServer.scala:377]: This is not great, so let me know if there's a better way to handle it (see selector.disconnected below to see where it's used).

**Comment by ijuma** [comment]
@junrao Fixed the issue you mentioned, tests passing locally. I also moved some comments I had mistakenly left on the commit into the "Files changed" tab to make the review easier.

**Comment by junrao** [comment]
Thanks for the patch. LGTM

**Pull request closed without merging**